File: golang/sliding_window/min_sub_array.go
package main

import "math"

func minSubArrayLen(target int, nums []int) int {
	n := len(nums)

	left, currentSum, minLength := 0, 0, math.MaxInt32

	for right := 0; right < n; right++ {
		currentSum += nums[right]

		for currentSum >= target {
			if minLength > right-left+1 {
				minLength = right - left + 1
			}

			currentSum -= nums[left]

			left++
		}
	}

	if minLength == math.MaxInt32 {
		return 0
	}

	return minLength
}

File: python/sliding_window/min_sub_array.py
from typing import List


class Solution:
    def minSubArrayLen(self, target: int, nums: List[int]) -> int | float:
        n = len(nums)

        left, currentSum, minLength = 0, 0, float("inf")

        for right in range(n):
            currentSum += nums[right]

            while currentSum >= target:
                minLength = min(minLength, right - left + 1)

                currentSum -= nums[left]

                left += 1

        return minLength if minLength != float("inf") else 0
